src/abstractdatatable/join.jl

```diff
@@ -162,31 +162,31 @@ DataTableJoiner{DT1<:AbstractDataTable, DT2<:AbstractDataTable}(dtl::DT1, dtr::D
     DataTableJoiner{DT1,DT2}(dtl, dtr, on)
 
 # helper map between the row indices in original and joined table
-immutable _RowIndexMap
+immutable RowIndexMap
     orig::Vector{Int} # row indices in the original table
     join::Vector{Int} # row indices in the resulting joined table
 end
 
-Base.length(x::_RowIndexMap) = length(x.orig)
+Base.length(x::RowIndexMap) = length(x.orig)
 
 # fix the result of the rightjoin by taking the nonnull values from the right table
 function fix_rightjoin_column!(res_col::AbstractArray, col_ix::Int, joiner::DataTableJoiner,
-                               all_orig_left_ixs::Vector{Int}, rightonly_ixs::_RowIndexMap)
+                               all_orig_left_ixs::Vector{Int}, rightonly_ixs::RowIndexMap)
     res_col[rightonly_ixs.join] = joiner.dtr_on[rightonly_ixs.orig, col_ix]
     res_col
 end
 
 # composes the joined data table using the maps between the left and right
 # table rows and the indices of rows in the result
 function compose_joined_table(joiner::DataTableJoiner,
-                left_ixs::_RowIndexMap, leftonly_ixs::_RowIndexMap,
-                right_ixs::_RowIndexMap, rightonly_ixs::_RowIndexMap)
+                              left_ixs::RowIndexMap, leftonly_ixs::RowIndexMap,
+                              right_ixs::RowIndexMap, rightonly_ixs::RowIndexMap)
     @assert length(left_ixs) == length(right_ixs)
     # compose left half of the result taking all left columns
     all_orig_left_ixs = vcat(left_ixs.orig, leftonly_ixs.orig)
```

Comment:
I think here we can avoid allocating one intermediate index array and simplify the code:
```julia
all_orig_left_ixs = similar(left_ixs.orig, length(left_ixs)+length(leftonly_ixs))
all_orig_left_ixs[left_ixs.join] = left_ixs.orig
(length(leftonly_ixs) > 0) && (all_orig_left_ixs[leftonly_ixs.join] = leftonly_ixs.orig)
```